Prevent instances shared memory segments clashes

ankor2023 opened the pull request:
Add uniq instance hash-codes generated from the instacne PID file name to the names of the shared memory segments and ipc-files.
This prevents shared memory clashes when running multiple instances with the same service name.
It is an implementation of the A-example, which was described in the mailing list:
https://lists.squid-cache.org/pipermail/squid-users/2025-March/027500.html

It uses murmur3_32 hash algorithm to produce an 8-character code from the filename of the PID of the instance and add it to the corresponding filenames (in functions Ipc::Mem::Segment::GenerateName(), Ipc::Port::MakeAddr() and Ipc::Port::CoordinatorAddr()).

Some examples of names:
```/dev/squid.e7815e09-cf__metadata.shm
/dev/squid.e7815e09-cf__queues.shm
/dev/squid.e7815e09-cf__readers.shm
/dev/squid.e7815e09-tls_session_cache.shm

/var/run/squid/squid.e7815e09-coordinator.ipc
/var/run/squid/squid.e7815e09-kid-1.ipc
/var/run/squid/squid.e7815e09-kid-2.ipc
/var/run/squid/squid.e7815e09-kid-3.ipc
```

yadij commented:
Build issues are because the `src/Instance.cc` is not linked against the unit tests for Rock and UFS cache types. See src/Makefile.am "tests for fs/*". Please run "make check" locally to check the fix for these and find any other tests impacted by this new code.


Also, why the murmur hash? The original proposal was just to add the PID number (from the "master" or "coordinator" process). If we do need to hash for any reason Squid uses [`libnettle`](https://www.lysator.liu.se/~nisse/nettle/nettle.html), for a selection of speed and security options.

rousskov commented on src/Instance.cc:
```diff
@@ -24,6 +24,53 @@
+// murmur3 hash function (32-bit)
+// Algorithm: https://en.wikipedia.org/wiki/MurmurHash
+static uint32_t
+murmur3_32(const char *key, size_t len, uint32_t seed) {
```
The use case behind this PR does not require a specific hash function. Please use a hash function that Squid already uses (instead of introducing a new one, especially a new one with an implementation embedded in Squid!).

Squid uses MD5, hash_string(), and hash4() (at least).

I suggest using MD5 because it should handle similar PID file names better than the other two functions. See storeKeyPublic() for an example of how to use MD5 in Squid.

ankor2023 replied:
Alex, thanks for the comment and for the reference to the MD5 usage example.
Using the example, I have converted the code:
```
SBuf
Instance::GetPidFilenameHash()
{
    cache_key hash[SQUID_MD5_DIGEST_LENGTH];
    const auto name = PidFilenameCalc().c_str();

    SquidMD5_CTX ctx;
    SquidMD5Init(&ctx);
    SquidMD5Update(&ctx, (unsigned char*) name, strlen(name));
    SquidMD5Final(hash, &ctx);

    SBuf pid_filename_hash;
    pid_filename_hash.appendf("%08x", *(int32_t*) hash);
    return pid_filename_hash;
}
```
Do you think 8 characters in a hash code (4 bytes from md5) in the instances naming is enough?
Or is it better to use an entire MD5 string?

rousskov replied:
> Do you think ... 4 bytes from md5 ... in the instances naming is enough?

Yes, I do. If deployment experience proves that more bytes are needed, we will increase the number of bytes.

If possible without heroic efforts, I would represent those 4 hash bytes in a more compact way than 8-byte hex encoding: The more stuff we add to these filenames, the higher are the chances we are going to hit some OS limits somewhere. Please note that I do not know whether this is possible (without heroic efforts).

> Or is it better to use an entire MD5 string?

No, using the whole MD5 hash here would be a (noisy and slightly dangerous) overkill IMO.

P.S. Please use C++ casts (e.g., reinterpret_cast) instead of C casts.

ankor2023 replied:
Hello Alex,

> f possible without heroic efforts, I would represent those 4 hash bytes in a more compact way than 8-byte hex encoding: The more stuff we add to these filenames, the higher are the chances we are going to hit some OS limits somewhere. Please note that I do not know whether this is possible (without heroic efforts).

We can use numbers and upper case letters to encode bytes more efficiently:
```
SBuf
Instance::GetPidFilenameHash()
{
    const auto code_table = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    cache_key hash[SQUID_MD5_DIGEST_LENGTH];
    unsigned char *p = hash;
    const auto name = PidFilenameCalc().c_str();

    SquidMD5_CTX ctx;
    SquidMD5Init(&ctx);
    SquidMD5Update(&ctx, (unsigned char*) name, strlen(name));
    SquidMD5Final(hash, &ctx);

    SBuf pid_filename_hash;
    pid_filename_hash.appendf("%c%c%c%c", code_table[p[0] % 36], code_table[p[1] % 36], code_table[p[2] % 36], code_table[p[3] % 36]);
    return pid_filename_hash;
}
```
This produces a 4-chars hash with 36 x 36 x 36 x 36 = 1679616 combinations.

ankor2023 replied:
Using Amos' recommendations, it will look like:

```
SBuf
Instance::GetPidFilenameHash()
{
    const auto code_table = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    cache_key hash[SQUID_MD5_DIGEST_LENGTH];
    unsigned char *p = hash;

    SquidMD5_CTX ctx;
    SquidMD5Init(&ctx);
    const auto name = PidFilenameCalc();
    SquidMD5Update(&ctx, name.rawContent(), name.length());
    SquidMD5Final(hash, &ctx);

    SBuf pid_filename_hash;
    pid_filename_hash.appendf("%c%c%c%c", code_table[p[0] % 36], code_table[p[1] % 36], code_table[p[2] % 36], code_table[p[3] % 36]);
    return pid_filename_hash;
}
```

rousskov commented on src/Instance.cc:
```diff
@@ -222,3 +269,15 @@ Instance::WriteOurPid()
+SBuf
+Instance::GetPidFilenameHash()
+{
+    SBuf name = PidFilenameCalc();
```
Please use const and [AAA](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/) whenever practical, including this line.

ankor2023 replied:
Alex, thank you for the comment.

> Please use `const` and [AAA](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/) whenever practical, including this line that can use AAA and, with a bit more effort, probably `const`.

I have a problem here.
For example, when I use the following program code:
```
SBuf
Instance::GetPidFilenameHash()
{
    const auto name = PidFilenameCalc();
    const auto name_c_str = name.c_str();
```
I get an error, the reason for which I cannot understand:
```
Instance.cc:278:40: error: passing ‘const SBuf’ as ‘this’ argument discards qualifiers [-fpermissive]
     const auto name_c_str = name.c_str();
                                        ^
```
The solution I foud is:
```
SBuf
Instance::GetPidFilenameHash()
{
    const auto name = PidFilenameCalc().c_str();
    const auto hash = murmur3_32(name, strlen(name), 0);
```
Is it OK?

rousskov replied:
> The solution I foud is:
> 
> ```C++
>     const auto name = PidFilenameCalc().c_str();
> ```
> 
> Is it OK?

It is a bug. Rule of thumb: Do not store c_str() result. The memory it points to will disappear in many use cases.

Please see Amos [response](https://github.com/squid-cache/squid/pull/2023#discussion_r1995462090) for the right approach to accessing SBuf content in this context (but use a different hash function as detailed in [another change request](https://github.com/squid-cache/squid/pull/2023#discussion_r1993542857), of course).

P.S. To reduce noise and CI costs, please test build locally and then push all your changes to the public repository _at once_ instead of abusing CI for trivial build testing and pushing every commit.

rousskov commented on src/ipc/Port.cc:
```diff
@@ -54,6 +55,8 @@ String Ipc::Port::MakeAddr(const char* processLabel, int id)
     assert(id >= 0);
     String addr = channelPathPfx;
     addr.append(service_name.c_str());
+    addr.append('.');
```
Please continue to use dashes (`-`) as separators (instead of switching to `.` that has special significance for filenames). Also, if you separate hash from service name (here), then separate it from process label (below) as well.

The same concerns apply to Ipc::Port::CoordinatorAddr() and Ipc::Mem::Segment::GenerateName(), of course.

rousskov reviewed (changes requested):
> Amos: The original proposal was just to add the PID number 

It was not. [Proposal A](https://lists.squid-cache.org/pipermail/squid-users/2025-March/027499.html) was to add a hash of the full PID file name.

Adding a PID number instead of a hash would solve the problem with segment clashes as well but would create a worse problem: Segments left by crashed Squid instances will accumulate instead of being removed (because each instance is likely to have a unique master PID value, even instances that have the same PID file name).

> Amos: If we do need to hash for any reason Squid uses [`libnettle`](https://www.lysator.liu.se/~nisse/nettle/nettle.html), for a selection of speed and security options.

I agree that this PR should not introduce a new hash function. However, please do not rely on libnettle for this PR either; essentially, ignore that libnettle exists. See [one of the change requests](https://github.com/squid-cache/squid/pull/2023#discussion_r1993542857) for a specific hash function recommendation.

ankor2023 commented:
Hello Amos,

Thank you for your help.

> Build issues are because the `src/Instance.cc` is not linked against the unit tests for Rock and UFS cache types. See src/Makefile.am "tests for fs/*". Please run "make check" locally to check the fix for these and find any other tests impacted by this new code.

I added **Instance.o** to the **tests_test\*_LDADD** where ipc/libipc.la included. But this entailed the need to add **debug.o** as well.
Now there is a problem with multiple definitions of some methods from the debug module, and I do not know how to fix it.
Can you please help me?

I also added the global variable **Chrooted** to the tests to make them use Instance.o. Is it OK?

rousskov commented on src/tests/testRock.cc:
```diff
@@ -40,6 +40,9 @@
  */
 
+/// Define Chrooted variable to use Instance.o
+bool Chrooted = false;
```
> I also added the global variable Chrooted to the tests to make them use Instance.o.

@yadij, there are several documented problems and a TODO in main.cc that suggest removing problematic chroot support from Squid. Would you support such a removal (in a dedicated PR, of course)?

yadij replied:
I can't answer that at this time, nor an a generic way. IIRC there was some controversial opinions on what to change vs remove, and also our downstream have some chroot related needs that complicate the situation.

yadij commented on src/tests/testCacheManager.cc:
```diff
@@ -20,6 +20,9 @@
  * test the CacheManager implementation
  */
 
+/// Define Chrooted variable to use Instance.o
+bool Chrooted = false;
```
> I also added the global variable **Chrooted** to the tests to make them use Instance.o. Is it OK?

IMO that is fine as a workaround for globals. The "correct" code is a much larger change which should not be pulled into the scope of this PR.

However each addition should be commented differently. Like so:
```suggestion
// XXX: for Instance.cc linkage. keep in sync with main.cc
bool Chrooted = false;
```

yadij commented on src/Instance.cc:
```diff
@@ -222,3 +223,21 @@ Instance::WriteOurPid()
+Instance::GetPidFilenameHash()
+{
+    cache_key hash[SQUID_MD5_DIGEST_LENGTH];
+    const auto name = PidFilenameCalc().c_str();
+
+    SquidMD5_CTX ctx;
+    SquidMD5Init(&ctx);
+    SquidMD5Update(&ctx, (unsigned char*) name, strlen(name));
```
The hash function does not need to care about c-string termination. Direct use of the `name` buffer is safe here.

```suggestion

    SquidMD5_CTX ctx;
    SquidMD5Init(&ctx);
    const auto name = PidFilenameCalc();
    SquidMD5Update(&ctx, name.rawContent(), name.length());
```

yadij commented on src/Instance.h:
```diff
@@ -30,6 +30,10 @@ void WriteOurPid();
 /// Throws if PID file maintenance is disabled.
 pid_t Other();
 
+/// Returns MD5 hash of the PID file name to make a uniq service name
+/// to use it in the shared memory segments naming
```
We do not need to disclose the implementation details like being MD5. Just stating its a hash is fine.
```suggestion
/// \returns hash of the PID file name to make a unique service name for this Squid instance
```

ankor2023 commented on src/Makefile.am:
```diff
@@ -1147,6 +1147,8 @@ nodist_tests_testRock_SOURCES = \
 	swap_log_op.cc
 tests_testRock_LDADD = \
+        Instance.o \
+        debug/debug.o \
```
Hello Amos,

Thank you for the help and for the recomendations.

I changed the Makefile.am according to your recommendations, but there are errors.
The output of the `make check`:
```
make[3]: 'tests/testSBufList' is up to date.
make[3]: 'tests/testString' is up to date.
/bin/sh ../libtool  --tag=CXX   --mode=link g++ -std=c++17 -Wall -Wextra -Wimplicit-fallthrough=5 -Wpointer-arith -Wwrite-strings -Wcomments -Wshadow -Wmissing-declarations -Woverloaded-virtual -Werror -pipe -D_REENTRANT -g -O2 -march=native -I.. -I../include -I../lib -I../src -I../include   -I../src  -ldl  -g -o tests/testRock  unlinkd.o AccessLogEntry.o tests/stub_CacheDigest.o tests/stub_CachePeer.o CollapsedForwarding.o ConfigOption.o ConfigParser.o ETag.o EventLoop.o FadingCounter.o tests/stub_HelperChildConfig.o HttpBody.o HttpHdrCc.o HttpHdrContRange.o HttpHdrRange.o HttpHdrSc.o HttpHdrScTarget.o HttpHeader.o HttpHeaderTools.o HttpReply.o tests/stub_HttpRequest.o LogTags.o MasterXaction.o MemBuf.o MemObject.o MemStore.o Notes.o Parsing.o tests/stub_Port.o RemovalPolicy.o RequestFlags.o ResolvedPeers.o tests/testRock.o StatCounters.o tests/stub_StatHist.o StoreFileSystem.o StoreIOState.o tests/testStoreSupport.o StoreSwapLogData.o StrList.o String.o Transients.o tests/stub_access_log.o tests/stub_cache_cf.o tests/stub_cache_manager.o cbdata.o tests/stub_client_db.o tests/stub_client_side.o tests/stub_client_side_request.o tests/stub_debug.o tests/stub_errorpage.o event.o fatal.o fd.o fde.o filemap.o tests/stub_fqdncache.o fs_io.o tests/stub_http.o tests/stub_icp.o int.o tests/stub_ipc.o tests/stub_ipcache.o tests/stub_libanyp.o tests/stub_libauth.o tests/stub_liberror.o tests/stub_libeui.o tests/stub_libformat.o tests/stub_libicmp.o tests/stub_libip.o tests/stub_liblog.o tests/stub_libmgr.o tests/stub_libsecurity.o mem_node.o tests/stub_mime.o tests/stub_neighbors.o tests/stub_pconn.o tests/stub_stat.o stmem.o store.o tests/stub_store_client.o store_io.o store_key_md5.o tests/stub_store_rebuild.o tests/stub_store_stats.o store_swapout.o tests/stub_tools.o wordlist.o test_tools.o globals.o Instance.o SquidMath.o hier_code.o swap_log_op.o http/libhttp.la parser/libparser.la libsquid.la comm/libcomm.la fs/libfs.la  repl/liblru.a DiskIO/libdiskio.la acl/libacls.la acl/libapi.la acl/libstate.la anyp/libanyp.la eui/libeui.la  ipc/libipc.la base/libbase.la mem/libmem.la store/libstore.la adaptation/libadaptation.la sbuf/libsbuf.la time/libtime.la ../lib/libmisccontainers.la ../lib/libmiscencoding.la ../lib/libmiscutil.la   -lcppunit  -lgnutls  ../compat/libcompatsquid.la  -lnettle  -lm -lresolv -lrt
libtool: link: g++ -std=c++17 -Wall -Wextra -Wimplicit-fallthrough=5 -Wpointer-arith -Wwrite-strings -Wcomments -Wshadow -Wmissing-declarations -Woverloaded-virtual -Werror -pipe -D_REENTRANT -g -O2 -march=native -I.. -I../include -I../lib -I../src -I../include -I../src -g -o tests/testRock unlinkd.o AccessLogEntry.o tests/stub_CacheDigest.o tests/stub_CachePeer.o CollapsedForwarding.o ConfigOption.o ConfigParser.o ETag.o EventLoop.o FadingCounter.o tests/stub_HelperChildConfig.o HttpBody.o HttpHdrCc.o HttpHdrContRange.o HttpHdrRange.o HttpHdrSc.o HttpHdrScTarget.o HttpHeader.o HttpHeaderTools.o HttpReply.o tests/stub_HttpRequest.o LogTags.o MasterXaction.o MemBuf.o MemObject.o MemStore.o Notes.o Parsing.o tests/stub_Port.o RemovalPolicy.o RequestFlags.o ResolvedPeers.o tests/testRock.o StatCounters.o tests/stub_StatHist.o StoreFileSystem.o StoreIOState.o tests/testStoreSupport.o StoreSwapLogData.o StrList.o String.o Transients.o tests/stub_access_log.o tests/stub_cache_cf.o tests/stub_cache_manager.o cbdata.o tests/stub_client_db.o tests/stub_client_side.o tests/stub_client_side_request.o tests/stub_debug.o tests/stub_errorpage.o event.o fatal.o fd.o fde.o filemap.o tests/stub_fqdncache.o fs_io.o tests/stub_http.o tests/stub_icp.o int.o tests/stub_ipc.o tests/stub_ipcache.o tests/stub_libanyp.o tests/stub_libauth.o tests/stub_liberror.o tests/stub_libeui.o tests/stub_libformat.o tests/stub_libicmp.o tests/stub_libip.o tests/stub_liblog.o tests/stub_libmgr.o tests/stub_libsecurity.o mem_node.o tests/stub_mime.o tests/stub_neighbors.o tests/stub_pconn.o tests/stub_stat.o stmem.o store.o tests/stub_store_client.o store_io.o store_key_md5.o tests/stub_store_rebuild.o tests/stub_store_stats.o store_swapout.o tests/stub_tools.o wordlist.o test_tools.o globals.o Instance.o SquidMath.o hier_code.o swap_log_op.o  -ldl http/.libs/libhttp.a parser/.libs/libparser.a ./.libs/libsquid.a comm/.libs/libcomm.a fs/.libs/libfs.a repl/liblru.a DiskIO/.libs/libdiskio.a -lpthread acl/.libs/libacls.a acl/.libs/libapi.a acl/.libs/libstate.a anyp/.libs/libanyp.a eui/.libs/libeui.a ipc/.libs/libipc.a base/.libs/libbase.a mem/.libs/libmem.a store/.libs/libstore.a adaptation/.libs/libadaptation.a sbuf/.libs/libsbuf.a time/.libs/libtime.a ../lib/.libs/libmisccontainers.a ../lib/.libs/libmiscencoding.a ../lib/.libs/libmiscutil.a -lcppunit -lgnutls ../compat/.libs/libcompatsquid.a -lnettle -lm -lresolv -lrt
Instance.o: In function `RemoveInstance()':
//data/bcscripter/src/development/squid/src/Instance.cc:221: undefined reference to `Debug::StopCacheLogUse()'
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:5616: tests/testRock] Error 1
make[3]: Leaving directory '/data/bcscripter/src/development/squid/src'
make[2]: *** [Makefile:6661: check-am] Error 2
make[2]: Leaving directory '/data/bcscripter/src/development/squid/src'
make[1]: *** [Makefile:6102: check-recursive] Error 1
make[1]: Leaving directory '/data/bcscripter/src/development/squid/src'
make: *** [Makefile:6664: check] Error 2

```

Could you please help me.

ankor2023 replied:
Amos, to make a progress, I added a dummy function `Debug::StopCacheLogUse()`  to the file tests/stub_debug.cc:
```
// Dummy function
void Debug::StopCacheLogUse(){}
```
Is it OK?

ankor2023 replied:
Eeventually, to deal with the problem, I added `#include "sbuf/SBuf.h"` into the file `Instance.h`.
Is it correct?

yadij replied:
IMO the `sbuf/SBuf.h` include is fine. If Alex objects then `sbuf/forward.h` is alternative but would require a lot of other files to include `sbuf/SBuf.h`.

yadij reviewed (dismissed):
Looks good enough to me.

rousskov reviewed (changes requested):
There are many little problems that should be fixed. I would prefer to commit the fixes directly to the PR branch than to do coding via change requests. @ankor2023, are you OK with me committing those fixes to your branch? You would be able to revert or discuss any commits you do not like, of course.

ankor2023 commented:
Hello Alex and Amos, 

Thank you for your time and for the help.

@rousskov 
> There are many little problems that should be fixed. I would prefer to commit the fixes directly to the PR branch than to do coding via change requests. @ankor2023, are you OK with me committing those fixes to your branch? You would be able to revert or discuss any commits you do not like, of course.
Yes, I sent you an invitation.

Also, I think the `-n` option is no longer required and can be removed from the program code as part of the PR.

yadij commented:
> Also, I think the `-n` option is no longer required and can be removed from the program code as part of the PR.

While the `-n` and service name it provides may now be removed from the `/dev/shm` paths. AFAIK it is still be necessary to determine which PID filename is the target instance of manual `-k` commands.

ankor2023 commented:
@rousskov : Hello Alex,

I sent you an invitation to my fork of Squid so you can commit to the brunch prevent-instance-shared-memory-segments-clashes directly . Haven't you seen it yet?

rousskov commented:
> you can commit to the brunch prevent-instance-shared-memory-segments-clashes directly

Thank you. I will find the time to make the necessary edits, but it will take me at least a few days. Please ping me if you do not see any changes by April 7, 2025.

rousskov commented:
Sorry, this is still on my to-do list. Updated ETA: April 28, 2025.

rousskov reviewed (approved):
@ankor2023, I am done with my changes. Please review the current PR, including:

* CONTRIBUTORS adjustment that assumes the two affected entries are yours. I am aware that this change breaks [one CI test](https://github.com/squid-cache/squid/actions/runs/14625928840/job/41037262311?pr=2023). I cannot reproduce that failure locally and am puzzled by it, but I hope to figure it out. For now, please ignore that CI test failure.
* Updated PR title and description that relax earlier uniqueness claim.
* New hashAt() lambda that encapsulates your byte-to-filename-character conversion.

FWIW, I have provided rationale behind all changes in the corresponding branch commits. If you have any questions, I am happy to discuss, but please check the corresponding commit messages first. If you would like to adjust anything futher, then please do so!

rousskov commented:
> Alex: CONTRIBUTORS adjustment that assumes the two affected entries are yours. I am aware that this change breaks [one CI test](https://github.com/squid-cache/squid/actions/runs/14625928840/job/41037262311?pr=2023). I cannot reproduce that failure locally and am puzzled by it, but I hope to figure it out. For now, please ignore that CI test failure.

@ankor2023, I figured it out: That CI test does not test the tip of this PR branch. Instead, it tests the so called "merge commit" created by GitHub (it can be fetched from this official repository as `refs/pull/2023/merge`; see [fetch_pr_refs()](https://github.com/squid-cache/squid/blob/76a642b613ddb0d4450930945703c68343f8d133/test-suite/test-sources.sh#L48)). All this is normal and is to be expected. However, GitHub creates that merge commit using name/email from your GitHub account. Since you have not updated your GitHub account information to match this PR CONTRIBUTORS changes, and merge commit comes _after_ PR branch commit [a096b0a](https://github.com/squid-cache/squid/pull/2023/commits/a096b0ad8d294233ad6f9a0e87b926c59031fa1c), our CI script extracts that old account name/email information from that merge commit and expects to find that entry in CONTRIBUTORS.

There are several ways to address this problem, but I will start with the simplest one (for me): Would you mind updating your GitHub settings to match the ones you want to see in CONTRIBUTORS? If that is not a good option (for you), then I will suggest more complex alternatives.

ankor2023 commented:
@rousskov : Hello, Alex, 
thank you very much for the code polishing, and for the datailed comments on the commits.

rousskov reviewed (dismissed):
> > Alex: Would you mind updating your GitHub settings to match the ones you want to see in CONTRIBUTORS?
 
> I think the problem was that I had the option **Keep my email addresses private** checked in my github profile. Now I have desabled it.

Yes, that explains GitHub behavior. Thank you for that adjustment. FWIW, you may revert that setting back after this PR merges.

> Now, I believe, the CI process should be completed successfully. But I don't know how to trigger the CI process without additional commits. Could you start it?

I triggered a rebuild by pushing a new commit (with a simple comment adjustment). CI tests are all green now. I am requesting a re-review by Amos because there were many changes since his earlier approval.

yadij commented on src/ipc/mem/Segment.cc:
```diff
@@ -277,8 +278,7 @@ Ipc::Mem::Segment::GenerateName(const char *id)
     } else {
-        name.append('/');
-        name.append(service_name.c_str());
+        name.append(Instance::NamePrefix("/").c_str());
```
`c_str()` performs a data copy, which is bad when we are already forced to do a data-copy to get the result into `String`.
```suggestion
        const auto prefix = Instance::NamePrefix("/");
        name.append(prefix.rawContent(), prefix.length());
```

yadij replied:
Yes, `SBuf::c_str()` should be avoided whenever possible.

I disagree with your "not good enough" - it works, does not add new bugs, improves the code, and meets the scope of this PR. IMHO, further improvement requiring larger rewrites is perfectionism that risks going out of scope and delays this bug fix which we need to get merged.

rousskov replied:
> Yes, `SBuf::c_str()` should be avoided whenever possible.

I disagree with the "possible" bit, but I will work on eliminating these c_str() calls from PR changes to make progress.

> I disagree with your "not good enough" - it works, does not add new bugs, improves the code, and meets the scope of this PR. IMHO, further improvement requiring larger rewrites is perfectionism that risks going out of scope and delays this bug fix which we need to get merged.

I disagree that the proposed replacement improves code. I have already detailed the reasons behind that assessment.

The problem this change request is targeting is indeed out of this PR scope. Insisting on a solution delays this PR.

yadij replied:
Done

yadij commented on src/ipc/mem/Segment.cc:
```diff
@@ -277,8 +278,7 @@ Ipc::Mem::Segment::GenerateName(const char *id)
     } else {
-        name.append('/');
-        name.append(service_name);
+        name.append(Instance::NamePrefix("/"));
```
@rousskov, I suspect the `if` branch of this logic needs a matching call to `Instance::NamePrefix(BasePath);` - with a related change to `BasePath` ensuring it has a trailing `/`.

yadij replied:
Thanks for the double-check. Accepting for now.

kinkie commented:
> > FAIL: tests/testRock
> 
> CI "make check" tests [failed](https://github.com/squid-cache/squid/actions/runs/16037337461/job/45266947677) on MacOS. Twice. I see no errors in CI log[1](#user-content-fn-1-f7bc42c2fbfe43439f6f1361a4457884) so testRock executable probably built OK but then failed when executed. We will need to address this regression before this PR can be merged. @ankor2023, can you help with that? Since the failure is MacOS-specific, addressing this may require running tests on MacOS locally or using your GitHub repository to run/debug MacOS tests using GitHub Actions...
> 
> ## Footnotes
> 1. If there are indeed no error details in CI logs, then that is another problem we should address, but addressing _that_ problem is outside this PR scope, of course. [↩](#user-content-fnref-1-f7bc42c2fbfe43439f6f1361a4457884)

I did some digging, the problem is:

```
quatro squid/master ‹prevent-instance-shared-memory-segments-clashes› » ./btlayer-00-default/squid-8.0.0-VCS/_build/sub/src/tests/testRock
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: StatHist.cc enumInit (not implemented).
SKIP: ip/libip.la setEmpty (not implemented).
SKIP: ip/libip.la Config (not implemented).
SKIP: lmgr/libmgr.la ActionParams (not implemented).
SKIP: ip/libip.la setEmpty (not implemented).
SKIP: ip/libip.la setEmpty (not implemented).
SKIP: ip/libip.la setEmpty (not implemented).
SKIP: ip/libip.la setEmpty (not implemented).
SKIP: ip/libip.la setEmpty (not implemented).
SKIP: ip/libip.la setEmpty (not implemented).
SKIP: ip/libip.la setEmpty (not implemented).
SKIP: ip/libip.la setEmpty (not implemented).
.SKIP: tools.cc UsingSmp (not implemented).
SKIP: Instance.cc NamePrefix (not implemented).
SKIP: tools.cc releaseServerSockets (not implemented).
SKIP: tools.cc leave_suid (not implemented).
stub time| FATAL: Ipc::Mem::Segment::create failed to shm_open(squid-STUB-0000-tr_rebuild_stats.shm): (63) File name too long

stub time| Squid Cache (Version 8.0.0-VCS): Terminated abnormally.
FATAL: tools.cc required for use of PrintRusage
```

I had already met this problem in the past. MacOS only supports shorter than POSIX shared memory segment names. man shm_open on MacOS says:

```
     [ENAMETOOLONG]     name exceeded the name size limit.  This is currently PSHMNAMLEN characters (defined in <sys/posix_shm.h>), but this may
                        change in the future.
```

.. and the header file says
```C
#define PSHMNAMLEN      31      /* maximum name segment length we bother with */
```

rousskov commented:
@kinkie, excellent analysis, thank you!

I pushed an untested temporary workaround (branch commit 5c28433a080b2cee64101f3a3ff97237657f7db2). If it or a similar workaround works, we will need to decide whether to merge this PR even though it introduces[^1] a MacOS regression OR address the problem by creating a function that replaces long segment names with short hashes (at least on MacOS).

[^1]: This phrasing assumes that segments (with default "squid" service name) work OK on MacOS with current master/v8 code, during "make check" and at runtime, but I have not verified that assumption. If you can run master/v8 Squid on MacOS, please try to start SMP Squid with a rock cache_dir and default cache_mem. If that works, try again but with `-n some-long-name-here-that-exceeds-PSHMNAMLEN` command-line parameter.

ankor2023 commented:
> @kinkie, excellent analysis, thank you! It would be nice to improve our code and CI to capture such FATAL messages (and other "make check" messages) in GitHub Actions tests.
> 
> I pushed an untested temporary workaround (branch commit [5c28433](https://github.com/squid-cache/squid/commit/5c28433a080b2cee64101f3a3ff97237657f7db2)). If it or a similar workaround works, we will need to decide whether to merge this PR even though it introduces[1](#user-content-fn-1-b0ba1dbb7c9c48fb6e317fc25476c591) a MacOS regression OR address the problem by creating a function that replaces long segment names with short hashes (at least on MacOS).
> 
> ## Footnotes
> 1. This phrasing assumes that segments (with default "squid" service name) work OK on MacOS with current master/v8 code, during "make check" and at runtime, but I have not verified that assumption. If you can run master/v8 Squid on MacOS, please try to start SMP Squid with a rock cache_dir and default cache_mem. If that works, try again but with `-n some-long-name-here-that-exceeds-PSHMNAMLEN` command-line parameter. [↩](#user-content-fnref-1-b0ba1dbb7c9c48fb6e317fc25476c591)

@rousskov : 
Hello, Alex, 

In this PR, we have added 4 additional hash-symbols + one '-' symbol.
Would it be better to shorten the segment names by 5 characters?
Shm-segments on a Linux host:
```/dev/shm/squid-CRS8-cache_mem_ex.shm
/dev/shm/squid-CRS8-cache_mem_map_anchors.shm
/dev/shm/squid-CRS8-cache_mem_map_filenos.shm
/dev/shm/squid-CRS8-cache_mem_map_slices.shm
/dev/shm/squid-CRS8-cache_mem_space.shm
/dev/shm/squid-CRS8-cf__metadata.shm
/dev/shm/squid-CRS8-cf__queues.shm
/dev/shm/squid-CRS8-cf__readers.shm
/dev/shm/squid-CRS8-squid-page-pool.shm
/dev/shm/squid-CRS8-transients_map_anchors.shm
/dev/shm/squid-CRS8-transients_map_filenos.shm
/dev/shm/squid-CRS8-transients_map_slices.shm
```
We can change:
```
cache_mem_map -> cache_mm
transients_map -> trans_map
tls_session_cache -> tls_ses_cache
```

rousskov commented:
```
Newer FreeBSD version for package ztrack:
To ignore this error set IGNORE_OSVERSION=yes
  - package: 1305000
  - running userland: 1304000
repository FreeBSD contains packages for wrong OS version
```

@kinkie, if restarting that [failed FreeBSD v13.4 staging test](https://github.com/squid-cache/squid/actions/runs/16147974049/job/45571738941) is not the answer, then please help getting past that package versioning error.

kinkie commented:
CI unblocked by PR #2112

yadij commented:
backport to v7 completed.
